src/task.c

```diff
@@ -640,72 +661,69 @@ task_enqueue(task_system *ts,
       }
    }
    platform_condvar_signal(&group->cv);
-   return platform_condvar_unlock(&group->cv);
+   return task_group_unlock(group);
 }
 
+/*
+ * Run a task if the number of waiting tasks is at least min_backlog.
+ */
 static platform_status
-task_group_perform_one(task_group *group, uint64 min_backlog)
+task_group_perform_one(task_group *group, uint64 queue_scale_percent)
 {
    platform_status rc;
    task           *assigned_task = NULL;
-   if (group->current_waiting_tasks <= min_backlog) {
+
+   /* We do the queue size comparison in this round-about way to avoid
+      integer overflow. */
+   if (queue_scale_percent
+       && 100 * group->current_waiting_tasks / queue_scale_percent
+             < group->bg.num_threads)
+   {
       return STATUS_TIMEDOUT;
    }
 
-   platform_thread_cleanup_push((void (*)(void *))platform_condvar_unlock,
-                                &group->cv);
-   rc = platform_condvar_lock(&group->cv);
+   rc = task_group_lock(group);
    if (!SUCCESS(rc)) {
-      goto out;
+      return rc;
    }
 
-   assigned_task = task_group_get_one_task(group);
+   assigned_task = task_group_get_next_task(group);
 
-out:
+   /* It is important to update the current_executing_tasks while
+      holding the lock. The reason is that, if we release the lock
+      before updating current_executing_tasks, then another thread
+      might observe that both the number of enqueued tasks and the
+      number of executing tasks are both 0, and hence that the system
+      is quiescent, even though it is not. */
```

Comment:
Minor, but .. could you please format the comment block to conform the style we follow elsewhere for multi-line comment blocks? Thanks!